### ISSUE TITLE
Demonstration for using pre-computed LD matrix

### ISSUE THREAD

**Issue by ZikunY** [getian107/SuSiEx#7]
Dear developer, 

 Congratulation on the SuSiEx method, it certainly fills the great needs of the fine-mapping community. I was trying to use SuSiEx based on precomputed LD matrix corresponding to the summary statistics. However, I received the error message such as, 

"* Please provide precomputed LD and frequency files or a reference panel for each GWAS using --ref-file"

The command line of SuSiEx is the following:
SuSiEx/bin/SuSiEx \
    --sst_file=data/Susiex/AMR_sumstats.txt,data/Susiex/EAS_sumstats.txt \
    --n_gwas=3000,3000 \
    --ld_file=data/Susiex/AMR_LD.txt,data/Susiex/EAS_LD.txt \
    --out_dir=Output/Susiex \
    --out_name=SuSiEx_AMR_EAS \
    --level=0.95 \
    --pval_thresh=1 \
    --chr=20 \
    --bp=$top,$bottom \
    --snp_col=2,2 \
    --chr_col=1,1 \
    --bp_col=3,3 \
    --a1_col=4,4 \
    --a2_col=5,5 \
    --eff_col=6,6 \
    --se_col=7,7 \
    --pval_col=9,9 \
     --mult-step=True \
    --keep-ambig=True \
    --threads=16

 The original SuSiE simply requires the Z-scores and the LD matrix for the computation. I wonder if SuSiEx can do the same. I also noticed that there is no demonstration for using precomputed LD matrix with SuSiEx. It would be great if there is a clear example for using the precomputed LD matrix for SuSiEx. 

Thank you!

**Comment by yorkklause** [comment]
Hi ZikunY,

Thank you for choosing our method.

We sincerely apologize for the inconvenience caused by this update during your analysis.

In this C++ version, we have implemented the use of a binary format LD file instead of a readable text file. The speed of software reading the binary format file is much faster than reading a readable text file.

Could you use the Python script,

SuSiEx/utilities/SuSiEx_LD.py

to re-calculate your LD again? This code will generate a binary LD file, a ref file, and an allele frequency file. SuSiEx requires these three files as input.

To use this Python script, please find a bash script example here,

SuSiEx/examples/cal_ld.sh



In our example, you can use the following commands to get the fine-mapping results.

#Calulate LD for EUR
python2 ../utilities/SuSiEx_LD.py \
	--ref_file=EUR \
	--ld_file=EUR \
	--chr=1 \
	--bp=7314654,8314677 \
	--plink=../utilities/plink \
	--maf=0.005
#Calulate LD for AFR
python2 ../utilities/SuSiEx_LD.py \
	--ref_file=AFR \
	--ld_file=AFR \
	--chr=1 \
	--bp=7314654,8314677 \
	--plink=../utilities/plink \
	--maf=0.005
#run SuSiEx
../bin/SuSiEx \
	--sst_file=EUR.sumstats.txt,AFR.sumstats.txt \
	--n_gwas=50000,50000 \
	--ref_file=EUR,AFR \
	--ld_file=EUR,AFR \
	--out_dir=./ \
	--out_name=SuSiEx.EUR.AFR.output.cs95 \
	--level=0.95 \
	--pval_thresh=1 \
	--chr=1 \
	--bp=7314654,8314677 \
	--snp_col=2,2 \
	--chr_col=1,1 \
	--bp_col=3,3 \
	--a1_col=4,4 \
	--a2_col=5,5 \
	--eff_col=6,6 \
	--se_col=7,7 \
	--pval_col=9,9 \
	--plink=../utilities/plink \
	--mult-step=True \
	--keep-ambig=True \
	--threads=16

I hope this example has helped you solve your problem. If it has not resolved the issue or if you have any additional questions, please don't hesitate to reach out.

My best

Kai Yuan

**Comment by ZikunY** [comment]
Hi Kai, 

 Thank you so much for this detailed answer, appreciate it!

 Yes, I actually used SuSiEx_LD.py to compute the LD based on the ref files and run SuSiEx after. Thanks for the demonstration here. 

 I totally understand the need for improving the loading speed of LD matrix and it could be quite burdensome if the region was really large or there were many regions. However, it would be flexible for users if there was an alternative option that the users can directly provide the pre-computed LD matrix for SuSiEx along with the Z-scores. It would be easier for users to use and improve the popularity of the method. 

Thanks again for the answer!

Best,
Zikun

**Comment by yorkklause** [comment]
Hi Zikun,

Thank you for your suggestion. We will make an effort to incorporate the LD + z-scores format into future updates of SuSiEx.

My best

Kai Yuan

**Comment by Jesson-mark** [comment]
Hi Kai, 

I have a similar question regarding to the precomputed LD and frequency. 

I used the datasets in your examples directory and supplied ld files to --ld_file parameter via `--ld_file=tests/EUR.ld,tests/AFR.ld`. 
Unfortunately, the SuSiEx program errored out with "* Please provide precomputed LD and frequency files or a reference panel for each GWAS using --ref-file". 

However, I cannot find the arguments that accept the precomputed **frequency file**. I wonder if there is a more flexible way to specify the precomputed LD and frequency files? Besides, is the `.bin` suffix of parameters passed to --ld_file argument required (`--ld_file=tests/EUR.ld,tests/AFR.ld` or `--ld_file=tests/EUR.ld.bin,tests/AFR.ld.bin`)?

One final question: I’m curious why the frequency is required in SuSiEx, given that frequency information is already available in the summary statistics and is not required in SuSiE package.

Best, 
Jie Wang

**Comment by Jesson-mark** [comment]
Hi Kai, 

I checked the source code of SuSiEx and found that SuSiEx will automatically add ".ld.bin", "_frq.frq" and "_ref.bim" to ld_file parameters. It might be helpful to include this additional step in the README so that other users can know how to prepare the precomputed LD and frequency files without needing to check the source code. It might be clearer to have three separate arguments that accept the absolute paths for each of the three different file types, though I understand this could be a bit cumbersome. 

Besides, I wonder why the "_ref.bim" file are required, because the information may be also included in the summary statistics file and sometimes the bim file are not easy to accuire. 

Finally, after I specified `--ld_file=tests/EUR,tests/AFR` arguments and prepared those files, SuSiEx still errored out with "* Please provide the directory and filename of PLINK using --plink". It seems that plink are still required even after I already provided the precomputed LD and frequency files. I'm a bit unclear about the specific role of plink in this case. 

Best,
Jie Wang

**Comment by yorkklause** [comment]
Hi @Jesson-mark ,

I'm sorry for the confusion.

SuSiEx requires pre-calculated **binary** format LD information as input (*ld.bin). **We strongly recommend using the script located in the "utilities" folder, utilities/SuSiEx_LD.py, to prepare your LD input data.**

The reason we ask for the "_ref.bim" file is that it helps us identify which allele corresponds to 0 and which corresponds to 1 in the LD matrix, ensuring the LD data aligns with the summary statistics files.

Please note that the current version of SuSiEx does not support plain text format LD matrices (i.e., ".ld" files). We require the binary version, which is optimized for faster data loading and can be easily generated using the utilities/SuSiEx_LD.py script.

If you already have the LD matrix in plain text format, you can convert it to binary format using the script utilities/convert2bin.cpp.
1. Compile the script
   g++ utilities/convert2bin.cpp -o utilities/convert2bin  
2. Convert your LD matrix
   cat $path_to_ld_file | ./utilities/convert2bin $path_to_binary_ld_file  

If you would like to use your pre-calculated LD matrix, please make sure the "_ref.bim" file aligns with the LD matrix. 

My best

Kai Yuan

**Comment by Jesson-mark** [comment]
Hi Kai, 

thank you so much for your detailed explanation! Now I understood the importance of "_ref.bim" file. 

As to my project, I already calculated covariates-adjusted LD matrix and saved into a ".bin" file with single-precision. I'm curious about the preparation of "_ref.bim" and "_frq.frq" files associated with LD matrix. 

I processed my dataset using plink2 that will preserve reference and alternative alleles, which is different from plink1.9 that will use minor allele as A1 thus losing reference and alternative alleles. 

So what I need to do is to ensure that alleles (A1 and A2) in "_ref.bim" and "_frq.frq" files are consistent with those used in my LD calculation, is that correct?

Best, 

Jie Wang
